openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            usage_key (UsageKey) - Usage key of the block that is to
+                be added to the given block_relations.
+        """
+
+        # Since block_relations is a DefaultDict, the usage_key will
+        # automatically be added by virtue of accessing it.
+        _ = block_relations[usage_key]
+
+
+class _BlockData(object):
+    """
+    Data structure to encapsulate collected data for a single block.
+    """
+    def __init__(self):
+        # Map of xblock field name to the field's value for this block.
+        # dict {string: any picklable type}
+        self._xblock_fields = {}
+
+        # Map of transformer name to the transformer's data for this
+        # block.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
+
+
+class BlockStructureBlockData(BlockStructure):
+    """
+    Subclass of BlockStructure that is responsible for managing block
+    and transformer data.
+    """
+    def __init__(self, root_block_usage_key):
+        super(BlockStructureBlockData, self).__init__(root_block_usage_key)
+
+        # Map of a block's usage key to its collected data, including
+        # its xBlock fields and block-specific transformer data.
+        # defaultdict {UsageKey: _BlockData}
+        self._block_data_map = defaultdict(_BlockData)
+
+        # Map of a transformer's name to its non-block-specific data.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
+
+    def get_xblock_field(self, usage_key, field_name, default=None):
+        """
+        Returns the collected value of the xBlock field for the
+        requested block for the requested field_name; returns default if
+        not found.
+
+        Arguments:
+            usage_key (UsageKey) - Usage key of the block whose xBlock
+                field is requested.
+
+            field_name (string) - The name of the field that is
+                requested.
+
+            default (any type) - The value to return if a field value is
+                not found.
+        """
+        block_data = self._block_data_map.get(usage_key)
+        return block_data._xblock_fields.get(field_name, default) if block_data else default
+
+    def get_transformer_data(self, transformer, key, default=None):
+        """
+        Returns the value associated with the given key from the given
+        transformer's data dictionary; returns default if not found.
+
+        Arguments:
+            transformer (BlockStructureTransformer) - The transformer
+                whose collected data is requested.
+
+            key (string) - A dictionary key to the transformer's data
+                that is requested.
+        """
+        return self._transformer_data.get(transformer.name(), {}).get(key, default)
```

Comment:
Actually, this is a little more complicated than I realized.  If we use regular dictionary access, it will add the transformer to `self._transformer_data`, which is probably (maybe?) not what we want.